Add auth middleware

diogo-botelho opened the pull request:
Added added validation for busker and event routes:
- ensureCorrectUserOrAdminForBuskers
- ensureUserOwnsBuskerAccount
- ensureBuskerOwnsEvent

Added resLocals.ts file and helper functions:
- createResLocalsBuskers
- createResLocalsEvents

Refactored buskerRoutes tests

Added event method getAllEventIdsByBuskerName
Refactored busker method getAllBuskerNamesByUserId

diogo-botelho committed:
Added validation functions.

Added ensureCorrectUserOrAdminForBuskers.
Added ensureUserOwnsBuskerAccount.
Added ensureBuskerOwnsEvent.

diogo-botelho committed:
Added helper functions for res.locals.

Added createResLocalsBuskers and createResLocalsEvents.

diogo-botelho committed:
Made buskerName unique

diogo-botelho committed:
Refactored getAllBuskerNamesByUserId.

Changed variable buskerName to busker for added clarity.

diogo-botelho committed:
Added additional validation to Busker and Event Routes.

diogo-botelho committed:
Refactored buskerRoutes tests.

Accounted for additional route validation.

diogo-botelho committed:
Removed outdated tests.

diogo-botelho committed:
Added getAllEventIdsByBuskerName method.

diogo-botelho committed:
Modified buskerNew schema.

Separated userId from data about busker.

diogo-botelho committed:
Refactored busker register model and route.

Separated userId from data about busker (buskerData).

diogo-botelho committed:
Refactored busker model and routes tests.

To account for changes in model and routes regarding register method/route.

diogo-botelho committed:
Refactored middleware auth.ts
Added docstrings.
Removed unused code.

diogo-botelho committed:
Cleaned up busk-app-schema.sql

diogo-botelho committed:
Refactored resLocals.ts

diogo-botelho committed:
Refactored auth middleware.

ensure UserOwnsBuskerAccount checks for buskerName in req.body, as well as req.params.

diogo-botelho committed:
Refactored busker and event models.

diogo-botelho committed:
REfactored busker routes.

Took out try catch in remove route.
Refactored building res.locals.buskers when creating new busker account.

diogo-botelho committed:
Refactored event Routes.

diogo-botelho committed:
Refactored busker and event routes tests.

Cleaned up code.
Changed format of req.body for post, update and remove routes.

samau3 commented on backend/src/routes/buskerRoutes.ts:
```diff
@@ -61,8 +65,11 @@ router.get(
  * Returns { buskerId, buskerName, category, description }
  *
  **/
+// POST localhost:3000/buskers/userId
+// GET localhost:3000/buskers/buskerName
```
Remove unnecessary comments

diogo-botelho committed:
Changed jest configuration to run with jest -i

diogo-botelho committed:
Cleaned up unsued code

diogo-botelho committed:
Cleaned up auth function correctUserOrAdmin.

Deprecated correctUserOrAdminForBuskers.
Updated correctUserOrAdmin to look for req.params and req.body.
Updated routes.

diogo-botelho committed:
Updated auth tests based on changes made to auth middleware.
